pytorch_lightning/metrics/functional/auroc.py

```diff
@@ -0,0 +1,168 @@
+        preds = preds.transpose(0, 1).reshape(n_classes, -1).transpose(0, 1)
+        target = target.flatten()
+    if mode == 'multi-label' and preds.ndim > 2:
+        n_classes = preds.shape[1]
+        preds = preds.transpose(0, 1).reshape(n_classes, -1).transpose(0, 1)
+        target = target.transpose(0, 1).reshape(n_classes, -1).transpose(0, 1)
+
+    return preds, target, mode
+
+
+def _auroc_compute(
+        preds: torch.Tensor,
+        target: torch.Tensor,
+        mode: str,
+        num_classes: Optional[int] = None,
+        pos_label: Optional[int] = None,
+        average: Optional[str] = 'macro',
+        max_fpr: Optional[float] = None,
+        sample_weights: Optional[Sequence] = None,
+) -> torch.Tensor:
+    # check max_fpr parameter
+    if max_fpr is not None:
+        if (not isinstance(max_fpr, float) and 0 < max_fpr <= 1):
+            raise ValueError(f"`max_fpr` should be a float in range (0, 1], got: {max_fpr}")
+
+        if LooseVersion(torch.__version__) < LooseVersion('1.6.0'):
+            raise RuntimeError('`max_fpr` argument requires `torch.bucketize` which'
+                               ' is not available below PyTorch version 1.6')
+
+        # max_fpr parameter is only support for binary
+        if mode != 'binary':
+            raise ValueError("Partial AUC computation not available in "
+                             "multilabel/multiclass setting, 'max_fpr' must be"
+                             f" set to `None`, received `{max_fpr=}`.")
+
+    if mode == 'binary':
+        num_classes = 1
+
+    if mode == 'multi-label':
+        # for multilabel we iteratively evaluate roc in a binary fashion
+        output = [roc(preds[:, i], target[:, i], num_classes=1, pos_label=1, sample_weights=sample_weights)
+                  for i in range(num_classes)]
+        fpr = [o[0] for o in output]
+        tpr = [o[1] for o in output]
+    else:
+        fpr, tpr, _ = roc(preds, target, num_classes, pos_label, sample_weights)
+
+    if max_fpr is None or max_fpr == 1:
+        if num_classes != 1:
+            # calculate auc scores per class
+            auc_scores = [auc(x, y) for x, y in zip(fpr, tpr)]
+
+            # calculate average
+            if average is None:
+                return auc_scores
+            elif average == 'macro':
+                return torch.mean(torch.stack(auc_scores))
+            elif average == 'weighted':
+                if mode == 'multi-label':
+                    support = torch.sum(target, dim=0)
+                else:
+                    support = torch.bincount(target.flatten(), minlength=num_classes)
+                return torch.sum(torch.stack(auc_scores) * support / support.sum())
+            else:
+                allowed_average = (None, 'macro', 'weighted')
+                raise ValueError('Argument `average` expected to be one of the following:'
+                                 f' {allowed_average} but got {average}')
```

Comment:
`if - if - if - None` pattern is better than `if - elif - elif` in Early Return Pattern

Reply:
I agree, have changed it now :]